core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -768,6 +770,19 @@ public PlanOptimizers(
                             // Must run before AddExchanges and after ReplicateSemiJoinInDelete
                             // to avoid temporarily having an invalid plan
                             new DetermineSemiJoinDistributionType(costComparator, taskCountEstimator)))));
+
+            builder.add(new IterativeOptimizer(
+                    ruleStats,
+                    statsCalculator,
+                    estimatedExchangesCostCalculator,
+                    ImmutableSet.<Rule<?>>builder()
+                            .addAll(pushIntoTableScanRulesExceptJoins)
+                            .add(new PushJoinIntoTableScan(metadata))
```

Comment:
```
// PushJoinIntoTableScan must run after ReorderJoins (and DetermineJoinDistributionType) otherwise a too early
// pushdown could prevent optimal plan from being selected.